device/receive.go

```diff
@@ -103,100 +117,117 @@ func (device *Device) RoutineReceiveIncoming(recv conn.ReceiveFunc) {
 
-			// check size
+				// check size
 
-			if len(packet) < MessageTransportSize {
-				continue
-			}
+				if len(packet) < MessageTransportSize {
+					continue
+				}
 
-			// lookup key pair
+				// lookup key pair
 
-			receiver := binary.LittleEndian.Uint32(
-				packet[MessageTransportOffsetReceiver:MessageTransportOffsetCounter],
-			)
-			value := device.indexTable.Lookup(receiver)
-			keypair := value.keypair
-			if keypair == nil {
-				continue
-			}
+				receiver := binary.LittleEndian.Uint32(
+					packet[MessageTransportOffsetReceiver:MessageTransportOffsetCounter],
+				)
+				value := device.indexTable.Lookup(receiver)
+				keypair := value.keypair
+				if keypair == nil {
+					continue
+				}
 
-			// check keypair expiry
+				// check keypair expiry
 
-			if keypair.created.Add(RejectAfterTime).Before(time.Now()) {
-				continue
-			}
+				if keypair.created.Add(RejectAfterTime).Before(time.Now()) {
+					continue
+				}
 
-			// create work element
-			peer := value.peer
-			elem := device.GetInboundElement()
-			elem.packet = packet
-			elem.buffer = buffer
-			elem.keypair = keypair
-			elem.endpoint = endpoint
-			elem.counter = 0
-			elem.Mutex = sync.Mutex{}
-			elem.Lock()
+				// create work element
+				peer := value.peer
+				elem := device.GetInboundElement()
+				elem.packet = packet
+				elem.buffer = buffsArrs[i]
+				elem.keypair = keypair
+				elem.endpoint = endpoints[i]
+				elem.counter = 0
+				elem.Mutex = sync.Mutex{}
+				elem.Lock()
+
+				elemsForPeer, ok := elemsByPeer[peer]
+				if !ok {
+					elemsForPeer = device.GetInboundElementsSlice()
+					elemsByPeer[peer] = elemsForPeer
+				}
+				*elemsForPeer = append(*elemsForPeer, elem)
+				buffsArrs[i] = nil
+				continue
 
-			// add to decryption queues
-			if peer.isRunning.Load() {
-				peer.queue.inbound.c <- elem
-				device.queue.decryption.c <- elem
-				buffer = device.GetMessageBuffer()
-			} else {
-				device.PutInboundElement(elem)
-			}
-			continue
+			// otherwise it is a fixed size & handshake related packet
 
-		// otherwise it is a fixed size & handshake related packet
+			case MessageInitiationType:
+				okay = len(packet) == MessageInitiationSize
 
-		case MessageInitiationType:
-			okay = len(packet) == MessageInitiationSize
+			case MessageResponseType:
+				okay = len(packet) == MessageResponseSize
 
-		case MessageResponseType:
-			okay = len(packet) == MessageResponseSize
+			case MessageCookieReplyType:
+				okay = len(packet) == MessageCookieReplySize
 
-		case MessageCookieReplyType:
-			okay = len(packet) == MessageCookieReplySize
+			default:
+				device.log.Verbosef("Received message with unknown type")
+			}
 
-		default:
-			device.log.Verbosef("Received message with unknown type")
+			if okay {
+				select {
+				case device.queue.handshake.c <- QueueHandshakeElement{
+					msgType:  msgType,
+					buffer:   buffsArrs[i],
+					packet:   packet,
+					endpoint: endpoints[i],
+				}:
+					buffsArrs[i] = nil
```

Comment:
Similar question raised here as in https://github.com/WireGuard/wireguard-go/pull/64/files#r1117372586

Reply:
Full link to referenced thread is https://github.com/WireGuard/wireguard-go/pull/64#discussion_r1117372586.

Updated in https://github.com/WireGuard/wireguard-go/pull/64/commits/4e768c410bd7da3bcd5d758a3f0704e485cb6c0a.